lib/ipfs/Ipfs.ts

```diff
@@ -0,0 +1,163 @@
+import Timeout from './Util/Timeout';
+
+const multihashes = require('multihashes');
+
+/**
+ * Class that implements the `ICas` interface by communicating with IPFS.
+ */
+export default class Ipfs implements ICas {
+  private fetch = nodeFetch;
+
+  public constructor (public uri: string, private fetchTimeoutInSeconds: number) { }
+
+  public async write (content: Buffer): Promise<string> {
+    // A string that is cryptographically impossible to repeat as the boundary string.
+    const multipartBoundaryString = crypto.randomBytes(32).toString('hex');
+
+    // An exmaple of multipart form data:
+    //
+    // --ABoundaryString
+    //
+    // Content of the first part.
+    // --ABoundaryString
+    // Content-Type: application/octet-stream
+    //
+    // Binary content of second part.
+    // --ABoundaryString--
+    const beginBoundary = Buffer.from(`--${multipartBoundaryString}\n`);
+    const firstPartContentType = Buffer.from(`Content-Type: application/octet-stream\n\n`);
+    const endBoundary = Buffer.from(`\n--${multipartBoundaryString}--`);
+    const requestBody = Buffer.concat([beginBoundary, firstPartContentType, content, endBoundary]);
+
+    const requestParameters = {
+      method: 'POST',
+      headers: { 'Content-Type': `multipart/form-data; boundary=${multipartBoundaryString}` },
+      body: requestBody
+    };
+
+    const addUrl = url.resolve(this.uri, '/api/v0/add'); // e.g. 'http://127.0.0.1:5001/api/v0/add'
+    const response = await this.fetch(addUrl, requestParameters);
+
+    if (response.status !== HttpStatus.OK) {
+      console.error(`IPFS write error response status: ${response.status}`);
+
+      if (response.body) {
+        const errorBody = await ReadableStream.readAll(response.body);
+        console.error(`IPFS write error body: ${errorBody}`);
+      }
+
+      throw new SidetreeError(IpfsErrorCode.IpfsFailedWritingContent, `Failed writing content of ${content.length} bytes.`);
+    }
+
+    const body = await ReadableStream.readAll(response.body);
+    const base58EncodedMultihashString = JSON.parse(body.toString()).Hash;
+
+    // Convert base58 to base64url multihash.
+    const multihashBuffer = multihashes.fromB58String(base58EncodedMultihashString);
+    const base64urlEncodedMultihash = base64url.encode(multihashBuffer);
+
+    console.log(`Wrote ${content.length} byte content as IPFS CID: ${base58EncodedMultihashString}, base64url ID: ${base64urlEncodedMultihash}`);
+    return base64urlEncodedMultihash;
+  }
+
+  public async read (base64urlEncodedMultihash: string, maxSizeInBytes: number): Promise<FetchResult> {
+    // Convert base64url to base58 multihash.
+    let base58EncodedMultihashString;
+    try {
+      const multihashBuffer = base64url.toBuffer(base64urlEncodedMultihash);
+      multihashes.validate(multihashBuffer);
+      base58EncodedMultihashString = multihashes.toB58String(multihashBuffer);
+    } catch {
+      return { code: FetchResultCode.InvalidHash };
```

Comment:
Should we log the `error` thrown w/ the `base64urlEncodedMultihash` to help with debugging?

Reply:
Will do.